assets/blocks/core-image/index.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { addFilter } from '@wordpress/hooks';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { Spinner } from '@wordpress/components';
+import { useEffect } from '@wordpress/element';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import './style.scss';
+
+/**
+ * Add image credit meta to core/image block attributes
+ */
+addFilter(
+	'blocks.registerBlockType',
+	'newspack-plugin/register-hook/core-image',
+	( settings, name ) => {
+		if ( name !== 'core/image' ) {
+			return settings;
+		}
+		return {
+			...settings,
+			attributes: {
+				...settings.attributes,
+				meta: {
+					type: 'object',
+					default: {},
+				},
+			},
+		};
+	}
+);
+
+/**
+ * Display spinner and lock post saving until meta attributes are added to block
+ */
+const AttributesLoader = props => {
+	const imageId = props.attributes.id;
+	const { lockPostSaving, unlockPostSaving } = useDispatch( 'core/editor' );
+	const { meta } = useSelect( select => select( 'core' ).getMedia( imageId ), [ imageId ] ) ?? {
+		meta: {},
+	};
+	useEffect( () => {
+		// Meta added, proceed
+		if ( Object.keys( meta ).length ) {
+			props.setAttributes( { meta } );
+			unlockPostSaving( 'attachment-meta-empty' );
+		} else {
+			lockPostSaving( 'attachment-meta-empty' );
+		}
+	}, [ Object.keys( meta ).length ] );
+
+	return (
+		<>
+			{ ! Object.keys( meta ).length && (
+				<div className="newspack-block__core-image-background">
+					<div className="newspack-block__core-image-spinner">
+						<Spinner />
+					</div>
+				</div>
+			) }
+		</>
+	);
+};
+
+/**
+ * Helper to parse credit meta
+ *
+ * @param {{}} credit Credit meta to parse
```

Comment:
Sidenote: using TypeScript is preferred to JSDoc comments. But no harm in adding those.

Reply:
Thanks for pointing this out. I've gone ahead and migrated all the code to TS (I'm a huge advocate). I hope thats okay.

Reply:
That's great!